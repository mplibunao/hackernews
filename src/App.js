import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { sortBy } from 'lodash';
import classNames from 'classnames';
import './App.css';

const DEFAULT_QUERY = 'redux';
const DEFAULT_PAGE = 0;
const DEFAULT_HPP = '20';
//const DEFAULT_TAGS = 'story';
const DEFAULT_TAGS = 'story';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search_by_date';
const PARAM_SEARCH = 'query=';
const TAG_SEARCH = 'tags=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';
// Query for comments of a particular story
// https://hn.algolia.com/api/v1/search_by_date?tags=comment,story_14743596&page=0&hitsPerPage=20
// story_xxxx = `story_${hits.objectID}`

// Object containing the different sort functions by category
const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse(),
};

const updateSearchTopStoriesState = (hits, page) => (prevState) => {
  // get values from prevState instead of this.state
  const { searchKey, results } = prevState;
  // && acts as a gate to stop result from accessing non-existent searchKey index
  const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : [];
  const updatedHits = [...oldHits, ...hits];

  // return value is the new state we're setting
  return {
    results: {
      ...results,
      // concat at the end of the map the latest results
      [searchKey]: { hits: updatedHits, page}
    },
    isLoading: false,
  };
};

//const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}`;

/*
// Remember filter takes in value/item as parameter
// However we can't access searchterm from the scope (state), which we need for the conditional statament of the filter
// So we have to pass searchTerm as a parameter but that would ruin our filter function
// So we create a function which accepts the searchTerm state value then return a regular filter function

function isSearched(searchTerm){
  returns function(item){
    return !searchTerm || item.title.includes(searchTerm);
  }
}
// - !searchTerm is used to prematurely stop the filter and avoid checking the 2nd condition if searchTerm is empty
// Instead, filter just passes the origin array to map (since it's false);
// This is to prevent comparing the list.title with the empty search term which could result in false and cut out the whole array

// - If searchTerm exist, then first condition is always false which makes the || check the 2nd condition
// - Remember, || goes left to right; stops if 1st condition is true; else check 2nd statement
*/ 

// Higher Order function
//const isSearched = (searchTerm) => (item) => !searchTerm || item.title.toLowerCase().includes(searchTerm.toLowerCase());


class App extends Component {
  constructor(props){
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      isLoading: false,
    };

    /*
      The function is bound to the class and thus becomes a class method
      Need to bind class methods in the constructor; Binds `this` to class;
      By default, `this` inside elements point to a function not the class;
      Problem seems to arise when you pass these functions down to child components (this === undefined)
      This is a better alternative to calling `this.onDismiss` inside an arrow function since an arrow function is made every
        render which pile up and force the garbage collection to clean them (expensive)
    */
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
  }

  needsToSearchTopStories(searchTerm){
    return !this.state.results[searchTerm];
  }

  onSearchSubmit(event){
    const { searchTerm } = this.state;
    this.setState({searchKey: searchTerm});
    // If false, we don't fetch data and we just rerender (because of setState) which loads the right data because of searchKey
    if (this.needsToSearchTopStories(searchTerm)){
      this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
    }
    
    event.preventDefault();
  }

  // Concatenate recently fetched hits with previous hits
  setSearchTopStories({hits, page}){
    // pass a higher order function which returns a callback responsible for altering the state 'safely'
    this.setState(updateSearchTopStoriesState(hits, page))
  }

  fetchSearchTopStories(searchTerm, page){
    this.setState({ isLoading: true});
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${TAG_SEARCH}${DEFAULT_TAGS}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(e => e);
  }

  // Make API Request after initial rendering of component
  componentDidMount(){
    const { searchTerm } = this.state;
    this.setState({searchKey: searchTerm});
    this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
  }

  onDismiss(id){
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];
    // Filter function
    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page}
      }
    });
  }

  // Pass synthetic event to class method to access the event payload
  onSearchChange(event){
    this.setState({searchTerm: event.target.value });
  }

  render() {
    const {
      searchTerm,
      results,
      searchKey,
      isLoading,
    } = this.state;

    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
        </div>
          <Table
            list={list}
            onDismiss={this.onDismiss}
          />
          <div className="interactions">
            <ButtonWithLoading
              isLoading={isLoading}
              onClick={()=> this.fetchSearchTopStories(searchKey, page+1)}>
              More
            </ButtonWithLoading>
          </div>  
      </div>
    );
  }
}


/*
  You can only use concise body functions if you only have one expression
  In this case, this element tree is considered as one
  If you assign the props values to variable inside the function, that's more than one expression
*/
class Search extends Component{

  render() {
    const {
      value,
      onChange,
      onSubmit,
      children
    } = this.props;
    
    return(
      <form onSubmit={onSubmit}>
        {children} <input
          type="text"
          onChange={onChange}
          // set the value so the element doesn't handle its own state (uncontrolled components)
          value={value}
        />
        <button type="submit">
          {children}
        </button>
      </form>
    );
  }
}

Search.PropTypes = {
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

class Table extends Component {
  constructor(props){
    super(props);

    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.onSort = this.onSort.bind(this);
  }

  /**
   * onSort
   * @param {String} sortKey - Category used in sorting results
   * When you change sort categories, the sortKey === sortKey will automatically return false thus sort normally
   * When it is the same you just flip the previous sortOrder (until you change categories)
   */
  onSort(sortKey){
    const isSortReverse = this.state.sortKey === sortKey && !this.state.isSortReverse;
    this.setState({ sortKey, isSortReverse });
  }

  render(){

    const { list, onDismiss } = this.props;
    const { sortKey, isSortReverse } = this.state;

    const sortedList = SORTS[sortKey](list);
    const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

    return (
      <div className="table">
        {/** Table Header **/}
        <div className="table-header">
          <span style={{ width: '40%'}}>
            <Sort
              sortKey={'TITLE'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Title
            </Sort>
          </span>
          <span style={{ width: '30%' }}>
            <Sort
              sortKey={'AUTHOR'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Author
            </Sort>
          </span>
          <span style={{ width: '10%' }}>
            <Sort
              sortKey={'COMMENTS'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Comments
            </Sort>
          </span>
          <span style={{ width: '10%' }}>
            <Sort
              sortKey={'POINTS'}
              onSort={this.onSort}
              activeSortKey={sortKey}
            >
              Points
            </Sort>
          </span>
          <span style={{ width: '10%'}}>
            Archive
          </span>
        </div>

        {/** Table Body **/}
        { reverseSortedList.map( item =>
          <div key={item.objectID} className="table-row">
            <span style={{ width: '40%' }}>
              <a href={item.url}>{item.title}</a>
            </span>
            <span style={{ width: '30%' }}>
              {item.author}
            </span>
            <span style={{ width: '10%' }}>
              {item.num_comments}
            </span>
            <span style={{ width: '10%' }}>
              {item.points}
            </span>
            <span style={{ width: '10%' }}>
              <Button
                onClick={()=> onDismiss(item.objectID)}
                className="button-inline"
                >
                Dismiss
              </Button>
            </span>
          </div>
        )}
      </div>
    );
  }
}

Table.PropTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired,
};


const Button = ({onClick, className, children})=>
  <button
    onClick={onClick}
    className={className}
    type="button"
  >
    {children}
  </button>

Button.PropTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
};

Button.defaultProps = {
  className: '',
};

// Loading component
const Loading = ()=>
  <i className="fa fa-spinner fa-spin"></i>

// Higher Order Component returning a functional stateless component
const withLoading = (Component) => ({ isLoading, ...rest }) =>
  isLoading ? <Loading /> : <Component {...rest} />

/**
 * Store the returned functional component ({ isLoading, ...rest }) in variable
 * which has closure to Component argument (Button)
 */
const ButtonWithLoading = withLoading(Button);

// Reuse the Button component
const Sort = ({
  onSort,
  sortKey,
  children,
  activeSortKey
}) => { 

  const sortClass = classNames(
    'button-inline',
    { 'button-active': sortKey === activeSortKey }
  );

  return (
    <Button
      onClick={ ()=> onSort(sortKey) }
      className={sortClass}
    >
      {children}
    </Button>
  );
}


export default App;

export {
  Button,
  Search,
  Table,
  Loading,
  Sort,
};