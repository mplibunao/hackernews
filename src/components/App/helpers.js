
/**
 *  Higher Order Function that returns a callback function passed to setState
 *  This callback function is in charge of altering the state 
 */
const updateSearchTopStoriesState = (hits, page) => (prevState) => {
  // get values from prevState instead of this.state
  const { searchKey, results } = prevState;
  // && acts as a gate to stop result from accessing non-existent searchKey index
  const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : [];
  const updatedHits = [...oldHits, ...hits];

  // return value is the new state we're setting
  return {
    results: {
      ...results,
      // concat at the end of the map the latest results
      [searchKey]: { hits: updatedHits, page}
    },
    isLoading: false,
  };
};

export { updateSearchTopStoriesState };