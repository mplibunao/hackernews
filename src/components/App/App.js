import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Search from '../Search/';
import Table from '../Table/';
import Button from '../Button/';
import Loading from '../Loading/';

import './App.css';
import { updateSearchTopStoriesState } from './helpers';
import {
  DEFAULT_QUERY,
  DEFAULT_PAGE,
  DEFAULT_HPP,
  DEFAULT_TAGS,

  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  TAG_SEARCH,
  PARAM_PAGE,
  PARAM_HPP,
} from '../../constants/';

//const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}`;

/**
 * withLoading
 * @param {Component} Component
 * Higher Order Component that returns a functional stateless component
 * Inner functional component breaksdown props into two parts: One for conditional conditional rendering, one for props
 * Depending on isLoading value, either renders Loading Component or Component passed in the outer function (closure)
 */
const withLoading = (Component) => ({ isLoading, ...rest }) =>
  isLoading ? <Loading /> : <Component {...rest} />

/**
 * ButtonWithLoading
 * Store the returned functional component ({ isLoading, ...rest }) in variable
 * which has closure to Component argument (Button)
 */
const ButtonWithLoading = withLoading(Button);

ButtonWithLoading.PropTypes = {
  isLoading: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
}

class App extends Component {
  constructor(props){
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      isLoading: false,
    };

    /**
     *  @ Binding Functions to the Class as Class methods
     *  - Need to bind class methods inside the constructor; Binds `this` to class;
     *  - By default, `this` inside elements point to a function not the class;
     *  - Problem arises when you pass these functions down to child components (this === undefined)
     *  - Better than onDismiss={ ()=> this.onDismiss() } on child component since arrow function is made every
     *    - render which pile up and force the garbage collection to clean them (expensive)
     */
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
  }

  // Make API Request after initial rendering of component
  componentDidMount(){
    const { searchTerm } = this.state;
    this.setState({searchKey: searchTerm});
    this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
  }

  needsToSearchTopStories(searchTerm){
    return !this.state.results[searchTerm];
  }

  onSearchSubmit(event){
    const { searchTerm } = this.state;
    this.setState({searchKey: searchTerm});
    // If false, we don't fetch data and we just rerender (because of setState) which loads the right data because of searchKey
    if (this.needsToSearchTopStories(searchTerm)){
      this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
    }
    
    event.preventDefault();
  }

  // Concatenate recently fetched hits with previous hits
  setSearchTopStories({hits, page}){
    // pass a higher order function which returns a callback responsible for altering the state 'safely'
    this.setState(updateSearchTopStoriesState(hits, page));
    console.log(this.state);
  }

  fetchSearchTopStories(searchTerm, page){
    this.setState({ isLoading: true});
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${TAG_SEARCH}${DEFAULT_TAGS}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(e => e);
  }

  onDismiss(id){
    const { searchKey, results } = this.state;
    const { hits, page } = results[searchKey];
    // Filter function
    const isNotId = item => item.objectID !== id;
    const updatedHits = hits.filter(isNotId);
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page}
      }
    });
  }

  // Pass synthetic event to class method to access the event payload
  onSearchChange(event){
    this.setState({searchTerm: event.target.value });
  }

  render() {
    const {
      searchTerm,
      results,
      searchKey,
      isLoading,
    } = this.state;

    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}
          >
            Search
          </Search>
        </div>
          <Table
            list={list}
            onDismiss={this.onDismiss}
          />
          <div className="interactions">
            <ButtonWithLoading
              isLoading={isLoading}
              onClick={()=> this.fetchSearchTopStories(searchKey, page+1)}>
              More
            </ButtonWithLoading>
          </div>  
      </div>
    );
  }
}

export default App;

export { withLoading, ButtonWithLoading  };