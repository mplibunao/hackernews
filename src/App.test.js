import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import App, { Search, Button, Table, Loading } from './App';

describe('App', ()=>{

  /**
   *  @ Default test for create-react-app
   *  Renders the component and its children to a div element
   */
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  /**
   *  @ Snapshot Test
   *  Makes a snapshot of your rendered component and runs this snapshot against future snapshots
   *  Notifies you of any changes and asks if you want to update the snapshot
   */
  test('snapshots', ()=>{
    const component = renderer.create(
      <App />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Search', ()=>{

  it('renders', ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<Search>Search</Search>, div);
  });

  test('snapshots', ()=>{
    const component = renderer.create(
      <Search>Search</Search>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Button', ()=>{

  it('renders', ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<Button>Give Me More</Button>, div);
  });

  test('snapshots', ()=>{
    const component = renderer.create(
      <Button>Give Me More</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Loading', ()=>{
  
  it('renders', ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<Loading />, div);
  });

  test('snapshots', ()=>{
    const component = renderer.create(
      <Loading />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Table', ()=>{

  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' },
    ]
  };

  /** 
   *  @ Unit Test 
   *  Use shallow to render component and assert that Table has two items
   *  - Shallow renders the component without child components. You can make the test dedicated to one component
   *  - Enzyme has two other rendering mechanisms: mount() and render()
   *  Both creates instances of the parent component and all child components
   *  However mount() gives you more access to the component lifecycle methods
   * 
   *  Rule of thumb:
   *  - Always begin with a shallow test
   *  - If componentDidMount() or componentDidUpdate() should be tested, use mount()
   *  - If you want to test component lifecycle and children behavior, use mount()
   *  - If you want to test children with less overhead than mount() and you are not interested in lifecycle methods, use render()
   */ 
  it('shows two items in list', ()=>{
    const element =  shallow(
      <Table {...props} />
    );

    expect(element.find('.table-row').length).toBe(2);
  });

  it('renders', ()=>{
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('snapshots', ()=>{
    const component = renderer.create(
      <Table {...props} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

