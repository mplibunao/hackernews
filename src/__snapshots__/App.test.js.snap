// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      Search
       
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
  <div
    className="table"
  >
    <div
      className="table-header"
    >
      <span
        style={
          Object {
            "width": "40%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Title
        </button>
      </span>
      <span
        style={
          Object {
            "width": "30%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Author
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Comments
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Points
        </button>
      </span>
      <span
        style={
          Object {
            "width": "10%",
          }
        }
      >
        Archive
      </span>
    </div>
  </div>
  <div
    className="interactions"
  >
    <i
      className="fa fa-spinner fa-spin"
    />
  </div>
</div>
`;

exports[`Button snapshots 1`] = `
<button
  className=""
  onClick={undefined}
  type="button"
>
  Give Me More
</button>
`;

exports[`Loading snapshots 1`] = `
<i
  className="fa fa-spinner fa-spin"
/>
`;

exports[`Search snapshots 1`] = `
<form
  onSubmit={undefined}
>
  Search
   
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Archive
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
